CORDA-3220: Updating FinalityFlow with functionality to specify StatesToRecord

rogersanick opened the pull request:
👮🏻👮🏻👮🏻 !!!! DESCRIBE YOUR CHANGES HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

Updating FinalityFlow with functionality to allow the initiator to indicate the appropriate StatesToRecord. This allows the initiating party to record states in transactions which they are proposing, but are not necessarily participants of. 

An example would be the issuance of a token where the only participant is the holder NOT the issuer.

JIRA Tickets - Corda-3220

# PR Checklist:

- [✅ ] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [✅ ] If you added public APIs, did you write the JavaDocs?
- [✅ ] If the changes are of interest to application developers, have you added them to the [changelog](https://github.com/corda/corda/blob/master/docs/source/changelog.rst), and potentially the [release notes](https://github.com/corda/corda/blob/master/docs/source/release-notes.rst)?
- [✅ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/corda/corda/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/corda/corda/blob/master/CONTRIBUTING.md#developer-certificate-of-origin).

Thanks for your code, it's appreciated! :)

rogersanick committed:
Updating FinalityFlow with functionality to allow the initiator to indicate the appropriate StatesToRecord. This allows the intiating party to record states in transactions which they are proposing but are not necessarily participants of. An example would be the issuance of a token where the only participant is the holder NOT the issuer.

JIRA Tickets - Corda-3220

rogersanick committed:
Adding param annotation with description of StatesToRecord for FinalityFlow

rogersanick committed:
Adding an update to the changelog to reflect modifications to FinalityFlow.

rogersanick commented:
- Closed old PR and reopened this one with a new branch name (nickRogersFinalityFlowFix). Branches prefaced with 'release' have specific admin permissions applied to them.
- Adjusted target branch to 4.4 (new branch cut)
- Added in feedback regarding kDoc updates.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -68,13 +69,15 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
-            progressTracker: ProgressTracker = tracker()
-    ) : this(transaction, emptyList(), progressTracker, sessions, true)
+            progressTracker: ProgressTracker = tracker(),
+            statesToRecord: StatesToRecord = ONLY_RELEVANT
```
The API check is failing:

```
[14:56:45]	[Step 7/14] Number of API removals/changes: 1
[14:56:45]	[Step 7/14] -  public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection, net.corda.core.utilities.ProgressTracker, int, kotlin.jvm.internal.DefaultConstructorMarker)
```

You wouldn't know what the issue is by looking at the error message. The c'tor that is missing is a "hidden" one that's needed by apps written in Kotlin. Whenever a method or c'tor has default values then a special overload is created with `kotlin.jvm.internal.DefaultConstructorMarker` at then end. It's an implementation detail of Kotiln's default values feature.

However the downside is that by adding a new parameter here we've deleted this c'tor (and created a new one which is exactly the same but with the new parameter).

TL;DR - What you need to do is leave this c'tor as is and create a new one with `statesToRecord`.

rogersanick replied:
Thanks @shamsasari! I just added the new constructor. Hopefully, this resolves the issue. 

I assumed that the @ JVMOverloads annotation would solve this, but it seems like it just generates conditionals that call with different default values in different circumstances.

rogersanick committed:
Fixing API Stability bug where constructor required by Kotlin apps was being removed as a result of the addition of a default parameter.

Fixed by adding an additional constructor requiring StatesToRecord to be passed in as an arguement.

rogersanick committed:
Removed default value from new parameter as it added no value and conflated method signatures.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -68,6 +69,7 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
      * @param transaction What to commit.
      * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
      * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
```
Delete

rogersanick replied:
Deleted! Thanks.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -76,6 +78,21 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+    constructor(
+            transaction: SignedTransaction,
+            sessions: Collection<FlowSession>,
+            progressTracker: ProgressTracker = tracker(),
```
Move this param to the end

rogersanick replied:
Assuming you're referring to the progressTracker param? So the order would be:

transaction, sessions, statesToRecord, progressTracker

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -76,6 +78,21 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
+     */
+    constructor(
```
You will need to add `@JvmOverloads` because of the default tracker value.

rogersanick replied:
Added!

rogersanick committed:
Multiple minor bug fixes:
- added JVMOverloads tag to handle progressTracker with default value
- reordered parameter to match pattern of other constructors
- removed erronious param tag

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -76,6 +77,22 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault. Defaults to [StatesToRecord.ONLY_RELEVANT].
```
There's no default on this c'tor. Instead you probably want to mention the default value in the kdocs for the other c'tors.

rogersanick replied:
Removed! Thanks, @shamsasari 

I've added a section in the initial comments explaining the impact of `StatesToRecord` on the initiating FinalityFlow. Is that sufficient? How can I indicate the default value in the kdocs for other c'tors (the @param annotation didn't seem appropriate given that it wasn't a param)

rogersanick committed:
- Added information to FinalityFlow explanation indicating to the developer the impacts of StatesToRecord for the initiator
- Removed erroronious param message suggesting a default value for StatesToRecord in the new constructor added in the last commit

rogersanick committed:
Default behaviour for initiating FinalityFlow is Only_Relevant not All_Visible as was written in the comments

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -76,6 +81,22 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which transactions to commit to the vault.
```
Which _states_

rogersanick replied:
Updated for both instances of the @statesToRecord param

rogersanick committed:
Adjusting code comments to reflect statesToRecord parameter governs which states are to be recorded to the vault NOT which transactions.
